File: scorecalc/scorecalc.ts
import { randomUUID } from "node:crypto";
import { api, APIError } from "encore.dev/api";
import { Topic, Subscription } from "encore.dev/pubsub";
import log from "encore.dev/log";
import { prisma } from "./prisma";
import { StructKeyspace, expireInSeconds } from "encore.dev/storage/cache";
import { cluster } from "../cache";

import {
  scoreCalcRequestedTopic,
  scoreCalcCompletedTopic,
  scoreCalcFailedTopic,
  type ScoreCalcRequested,
  type ScoreCalcCompleted,
  type ScoreCalcFailed,
  type ScoreCalcProjection,
} from "./events";

interface DummyEventDetail {
  id: string;
}

// Duplicate keyspace definition to invalidate event-detail cache across service boundary
const eventDetailCache = new StructKeyspace<{ id: string }, DummyEventDetail>(cluster, {
  keyPattern: "event-detail/:id",
  defaultExpiry: expireInSeconds(120),
});

// Pub/sub topic to report score calculation status.
export interface ScoreCalcStatusEvent {
  eventId: string;
  userIds: string[];
  status: "queued" | "completed" | "failed";
  timestamp: string;
}

export const scoreCalcStatusTopic = new Topic<ScoreCalcStatusEvent>("score-calc-status", {
  deliveryGuarantee: "at-least-once",
});

export interface SubmitCalcParams {
  eventId: string;
  userIds: string[];
}

export interface SubmitCalcResponse {
  jobId: string;
  generation: number;
}

// Submits a request for score calculation for an event.
// Increments generation and publishes ScoreCalcRequested.
export const submitCalc = api(
  { method: "POST", expose: false },
  async (params: SubmitCalcParams): Promise<SubmitCalcResponse> => {
    if (params.userIds.length === 0) {
      throw APIError.invalidArgument("userIds list cannot be empty");
    }

    // 1. Validate event and check scoringType
    const event = await prisma.event.findUnique({
      where: { id: params.eventId },
      select: { id: true, scoringType: true },
    });

    if (!event) {
      throw APIError.notFound(`Event ${params.eventId} not found`);
    }

    if (event.scoringType !== 1) {
      log.info(`Defensively rejecting submitCalc for non-points event ${params.eventId} (scoringType=${event.scoringType})`);
      return { jobId: "no-op", generation: 0 };
    }

    // 2. Coalesce/increment generation and create pending job in transaction
    const { jobId, generation } = await prisma.$transaction(async (tx) => {
      const state = await tx.scoreCalcState.upsert({
        where: { eventId: params.eventId },
        create: {
          eventId: params.eventId,
          latestGeneration: 1,
          acceptedGeneration: 0,
        },
        update: {
          latestGeneration: { increment: 1 },
        },
      });

      const nextGen = state.latestGeneration;
      const id = randomUUID();

      await tx.scoreCalcJob.create({
        data: {
          id,
          eventId: params.eventId,
          generation: nextGen,
          requestedUserIds: params.userIds,
          status: "PENDING",
          attempts: 0,
          requestedAt: new Date(),
        },
      });

      return { jobId: id, generation: nextGen };
    });

    // 3. Publish request event to scorecalc-requested topic
    await scoreCalcRequestedTopic.publish({
      version: 1,
      jobId,
      eventId: params.eventId,
      generation,
      requestedAt: new Date().toISOString(),
    });

    // 4. Report status to public topic
    await scoreCalcStatusTopic.publish({
      eventId: params.eventId,
      userIds: params.userIds,
      status: "queued",
      timestamp: new Date().toISOString(),
    });

    return { jobId, generation };
  }
);

// Subscription handler for completed calculations
export async function handleScoreCalcCompleted(event: ScoreCalcCompleted): Promise<void> {
  const { jobId, eventId, generation, result, resultChecksum } = event;

  log.info(`Received ScoreCalcCompleted message: jobId=${jobId}, eventId=${eventId}, gen=${generation}`);

  try {
    const processed = await prisma.$transaction(async (tx) => {
      const job = await tx.scoreCalcJob.findUnique({
        where: { id: jobId },
      });

      if (!job) {
        log.warn(`Job ${jobId} not found in database. Skipping.`);
        return { success: false };
      }

      if (job.status === "COMPLETED" || job.status === "SUPERSEDED") {
        log.info(`Job ${jobId} is already ${job.status}. Skipping.`);
        return { success: false };
      }

      const state = await tx.scoreCalcState.findUnique({
        where: { eventId },
      });

      if (!state) {
        log.error(`ScoreCalcState not found for event ${eventId}. Skipping.`);
        return { success: false };
      }

      // Fencing check: stale calculation cannot overwrite a newer generation
      if (generation < state.latestGeneration) {
        log.info(`Ignoring stale completion: generation ${generation} is older than latest ${state.latestGeneration}.`);
        await tx.scoreCalcJob.update({
          where: { id: jobId },
          data: { status: "SUPERSEDED" },
        });
        return { success: false };
      }

      // Apply results and update job status
      for (const entry of result.entries) {
        await tx.eventPointsEntry.upsert({
          where: { eventId_userId: { eventId, userId: entry.userId } },
          create: {
            id: randomUUID(),
            eventId,
            userId: entry.userId,
            points: entry.points,
            resultStatus: entry.resultStatus ?? null,
          },
          update: {
            points: entry.points,
            resultStatus: entry.resultStatus ?? null,
          },
        });
      }

      // Update state and job atomically
      await tx.scoreCalcState.update({
        where: { eventId },
        data: { acceptedGeneration: generation },
      });

      await tx.scoreCalcJob.update({
        where: { id: jobId },
        data: {
          status: "COMPLETED",
          completedAt: new Date(),
          resultChecksum,
        },
      });

      const requestedUserIds = job.requestedUserIds as string[] | null;
      return { success: true, userIds: requestedUserIds ?? [] };
    });

    if (processed.success) {
      // Post-commit cache invalidation
      await eventDetailCache.delete({ id: eventId });

      // Publish status event with originally requested userIds
      await scoreCalcStatusTopic.publish({
        eventId,
        userIds: processed.userIds,
        status: "completed",
        timestamp: new Date().toISOString(),
      });
      log.info(`Successfully persisted score projection for jobId=${jobId}`);
    }

  } catch (err: any) {
    log.error(err, `Failed to handle ScoreCalcCompleted for Job ${jobId}`);
    throw err;
  }
}

// Subscription handler for failed calculations
export async function handleScoreCalcFailed(event: ScoreCalcFailed): Promise<void> {
  const { jobId, eventId, generation, errorMessage } = event;

  log.info(`Received ScoreCalcFailed message: jobId=${jobId}, eventId=${eventId}, gen=${generation}`);

  try {
    const processed = await prisma.$transaction(async (tx) => {
      const job = await tx.scoreCalcJob.findUnique({
        where: { id: jobId },
      });

      if (!job) {
        log.warn(`Job ${jobId} not found. Skipping.`);
        return { success: false };
      }

      if (job.status === "COMPLETED" || job.status === "SUPERSEDED") {
        return { success: false };
      }

      const state = await tx.scoreCalcState.findUnique({
        where: { eventId },
      });

      // Ensure stale failures don't regress newer generation's status
      if (state && generation < state.latestGeneration) {
        await tx.scoreCalcJob.update({
          where: { id: jobId },
          data: { status: "SUPERSEDED" },
        });
        return { success: false };
      }

      await tx.scoreCalcJob.update({
        where: { id: jobId },
        data: {
          status: "FAILED",
          lastError: errorMessage,
          completedAt: new Date(),
        },
      });

      const requestedUserIds = job.requestedUserIds as string[] | null;
      return { success: true, userIds: requestedUserIds ?? [] };
    });

    if (processed.success) {
      await scoreCalcStatusTopic.publish({
        eventId,
        userIds: processed.userIds,
        status: "failed",
        timestamp: new Date().toISOString(),
      });
    }

  } catch (err: any) {
    log.error(err, `Failed to handle ScoreCalcFailed for Job ${jobId}`);
    throw err;
  }
}

const _onCompleted = new Subscription(scoreCalcCompletedTopic, "scorecalc-completed-handler", {
  handler: handleScoreCalcCompleted,
});

const _onFailed = new Subscription(scoreCalcFailedTopic, "scorecalc-failed-handler", {
  handler: handleScoreCalcFailed,
});

File: scorecalc/scorecalc.go
package scorecalc

import (
	"context"
	"crypto/rand"
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"time"

	"encore.dev/beta/errs"
	"encore.dev/rlog"
	"encore.dev/storage/cache"
	"encore.dev/pubsub"
	"encore.app/shared"
)

// newID generates a random UUIDv4 string for row ids.
func newID() string {
	var b [16]byte
	_, _ = rand.Read(b[:])
	b[6] = (b[6] & 0x0f) | 0x40
	b[8] = (b[8] & 0x3f) | 0x80
	return fmt.Sprintf("%x-%x-%x-%x-%x", b[0:4], b[4:6], b[6:8], b[8:10], b[10:16])
}

// db is the shared Lightwing database (no aliasing: Encore tracks resources
// through direct package-level references only).
// eventDetailCache mirrors the TS duplicate keyspace definition: it exists
// so this service can invalidate the event-detail cache across the service
// boundary (same Redis keys as eventmanager's keyspace).
type eventDetailKey struct {
	ID string
}

type eventDetailValue struct {
	ID string `json:"id"`
}

var eventDetailCache = cache.NewStructKeyspace[eventDetailKey, eventDetailValue](shared.Cache, cache.KeyspaceConfig{
	KeyPattern:    "event-detail/:ID",
	DefaultExpiry: cache.ExpireIn(120 * time.Second),
})

var (
	_ = pubsub.NewSubscription(ScoreCalcCompletedTopic, "scorecalc-completed-handler", pubsub.SubscriptionConfig[ScoreCalcCompleted]{
		Handler: HandleScoreCalcCompleted,
	})
	_ = pubsub.NewSubscription(ScoreCalcFailedTopic, "scorecalc-failed-handler", pubsub.SubscriptionConfig[ScoreCalcFailed]{
		Handler: HandleScoreCalcFailed,
	})
)

// SubmitCalcParams requests a score calculation for an event.
type SubmitCalcParams struct {
	EventID string   `json:"eventId"`
	UserIDs []string `json:"userIds"`
}

// SubmitCalcResponse carries the queued job id and generation.
type SubmitCalcResponse struct {
	JobID      string `json:"jobId"`
	Generation int    `json:"generation"`
}

// SubmitCalc queues a score calculation: validates the event, coalesces the
// generation, persists a pending job, and publishes the request + status.
//
// Mirrors ts-legacy/scorecalc/scorecalc.ts submitCalc. Private: called from
// inside the app (eventmanager) only.
//
//encore:api private method=POST path=/scorecalc/submit
func SubmitCalc(ctx context.Context, params *SubmitCalcParams) (*SubmitCalcResponse, error) {
	if len(params.UserIDs) == 0 {
		return nil, &errs.Error{Code: errs.InvalidArgument, Message: "userIds list cannot be empty"}
	}

	var scoringType int
	err := db.QueryRow(ctx,
		`SELECT "scoringType" FROM "event" WHERE id = $1`, params.EventID,
	).Scan(&scoringType)
	if errors.Is(err, sql.ErrNoRows) {
		return nil, &errs.Error{Code: errs.NotFound, Message: fmt.Sprintf("Event %s not found", params.EventID)}
	}
	if err != nil {
		return nil, err
	}

	if scoringType != 1 {
		rlog.Info(fmt.Sprintf("Defensively rejecting submitCalc for non-points event %s (scoringType=%d)", params.EventID, scoringType))
		return &SubmitCalcResponse{JobID: "no-op", Generation: 0}, nil
	}

	userIDsJSON, err := json.Marshal(params.UserIDs)
	if err != nil {
		return nil, fmt.Errorf("failed to encode user ids: %w", err)
	}

	tx, err := db.Begin(ctx)
	if err != nil {
		return nil, fmt.Errorf("failed to begin transaction: %w", err)
	}
	defer tx.Rollback()

	var generation int
	err = tx.QueryRow(ctx,
		`INSERT INTO "score_calc_state" ("eventId", "latestGeneration", "acceptedGeneration", "updatedAt")
		 VALUES ($1, 1, 0, $2)
		 ON CONFLICT ("eventId") DO UPDATE SET "latestGeneration" = "score_calc_state"."latestGeneration" + 1
		 RETURNING "latestGeneration"`,
		params.EventID, time.Now().UTC(),
	).Scan(&generation)
	if err != nil {
		return nil, fmt.Errorf("failed to coalesce generation: %w", err)
	}

	jobID := newID()
	now := time.Now().UTC()
	_, err = tx.Exec(ctx,
		`INSERT INTO "score_calc_task" (id, "eventId", generation, "requestedUserIds", status, attempts, "requestedAt")
		 VALUES ($1, $2, $3, $4, 'PENDING', 0, $5)`,
		jobID, params.EventID, generation, string(userIDsJSON), now,
	)
	if err != nil {
		return nil, fmt.Errorf("failed to create job: %w", err)
	}
	if err := tx.Commit(); err != nil {
		return nil, fmt.Errorf("failed to commit job: %w", err)
	}

	if _, err := ScoreCalcRequestedTopic.Publish(ctx, ScoreCalcRequested{
		Version: 1, JobID: jobID, EventID: params.EventID,
		Generation: generation, RequestedAt: now.Format(time.RFC3339Nano),
	}); err != nil {
		return nil, fmt.Errorf("failed to publish request: %w", err)
	}
	if _, err := ScoreCalcStatusTopic.Publish(ctx, ScoreCalcStatusEvent{
		EventID: params.EventID, UserIDs: params.UserIDs,
		Status: "queued", Timestamp: now.Format(time.RFC3339Nano),
	}); err != nil {
		return nil, fmt.Errorf("failed to publish status: %w", err)
	}

	return &SubmitCalcResponse{JobID: jobID, Generation: generation}, nil
}

// scoreCalcJobRow mirrors the score_calc_task columns read by handlers.
type scoreCalcJobRow struct {
	ID               string
	EventID          string
	Generation       int
	RequestedUserIDs []byte
	Status           string
	Attempts         int
}

// HandleScoreCalcCompleted persists a finished projection with generation
// fencing: stale completions are marked SUPERSEDED and never overwrite newer
// generations.
//
// Mirrors ts-legacy/scorecalc/scorecalc.ts handleScoreCalcCompleted.
func HandleScoreCalcCompleted(ctx context.Context, event ScoreCalcCompleted) error {
	rlog.Info(fmt.Sprintf("Received ScoreCalcCompleted message: jobId=%s, eventId=%s, gen=%d", event.JobID, event.EventID, event.Generation))

	tx, err := db.Begin(ctx)
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %w", err)
	}
	defer tx.Rollback()

	var job scoreCalcJobRow
	err = tx.QueryRow(ctx,
		`SELECT id, "eventId", generation, "requestedUserIds", status, attempts
		 FROM "score_calc_task" WHERE id = $1`, event.JobID,
	).Scan(&job.ID, &job.EventID, &job.Generation, &job.RequestedUserIDs, &job.Status, &job.Attempts)
	if errors.Is(err, sql.ErrNoRows) {
		rlog.Warn(fmt.Sprintf("Job %s not found in database. Skipping.", event.JobID))
		return nil
	}
	if err != nil {
		return err
	}
	if job.Status == "COMPLETED" || job.Status == "SUPERSEDED" {
		rlog.Info(fmt.Sprintf("Job %s is already %s. Skipping.", event.JobID, job.Status))
		return nil
	}

	var latestGeneration int
	err = tx.QueryRow(ctx,
		`SELECT "latestGeneration" FROM "score_calc_state" WHERE "eventId" = $1`, event.EventID,
	).Scan(&latestGeneration)
	if errors.Is(err, sql.ErrNoRows) {
		rlog.Error(fmt.Sprintf("ScoreCalcState not found for event %s. Skipping.", event.EventID))
		return nil
	}
	if err != nil {
		return err
	}

	if event.Generation < latestGeneration {
		rlog.Info(fmt.Sprintf("Ignoring stale completion: generation %d is older than latest %d.", event.Generation, latestGeneration))
		_, err = tx.Exec(ctx, `UPDATE "score_calc_task" SET status = 'SUPERSEDED' WHERE id = $1`, event.JobID)
		if err != nil {
			return err
		}
		return tx.Commit()
	}

	now := time.Now().UTC()
	for _, entry := range event.Result.Entries {
		_, err = tx.Exec(ctx,
			`INSERT INTO "event_points_entry" (id, "eventId", "userId", points, "createdAt", "updatedAt")
			 VALUES ($1, $2, $3, $4, $5, $5)
			 ON CONFLICT ("eventId", "userId") DO UPDATE SET points = EXCLUDED.points`,
			newID(), event.EventID, entry.UserID, entry.Points, now,
		)
		if err != nil {
			return fmt.Errorf("failed to upsert points entry: %w", err)
		}
	}
	if _, err := tx.Exec(ctx,
		`UPDATE "score_calc_state" SET "acceptedGeneration" = $1 WHERE "eventId" = $2`,
		event.Generation, event.EventID,
	); err != nil {
		return err
	}
	if _, err := tx.Exec(ctx,
		`UPDATE "score_calc_task" SET status = 'COMPLETED', "completedAt" = $1, "resultChecksum" = $2 WHERE id = $3`,
		now, event.ResultChecksum, event.JobID,
	); err != nil {
		return err
	}

	var userIDs []string
	if len(job.RequestedUserIDs) > 0 {
		_ = json.Unmarshal(job.RequestedUserIDs, &userIDs)
	}
	if err := tx.Commit(); err != nil {
		return err
	}

	// Post-commit cache invalidation + status publish.
	if eventDetailCache != nil {
		_, _ = eventDetailCache.Delete(ctx, eventDetailKey{ID: event.EventID})
	}
	if _, err := ScoreCalcStatusTopic.Publish(ctx, ScoreCalcStatusEvent{
		EventID: event.EventID, UserIDs: userIDs,
		Status: "completed", Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
	}); err != nil {
		return err
	}
	rlog.Info(fmt.Sprintf("Successfully persisted score projection for jobId=%s", event.JobID))
	return nil
}

// HandleScoreCalcFailed marks a job failed without regressing newer
// generations to a stale failure.
//
// Mirrors ts-legacy/scorecalc/scorecalc.ts handleScoreCalcFailed.
func HandleScoreCalcFailed(ctx context.Context, event ScoreCalcFailed) error {
	rlog.Info(fmt.Sprintf("Received ScoreCalcFailed message: jobId=%s, eventId=%s, gen=%d", event.JobID, event.EventID, event.Generation))

	tx, err := db.Begin(ctx)
	if err != nil {
		return fmt.Errorf("failed to begin transaction: %w", err)
	}
	defer tx.Rollback()

	var job scoreCalcJobRow
	err = tx.QueryRow(ctx,
		`SELECT id, "eventId", generation, "requestedUserIds", status, attempts
		 FROM "score_calc_task" WHERE id = $1`, event.JobID,
	).Scan(&job.ID, &job.EventID, &job.Generation, &job.RequestedUserIDs, &job.Status, &job.Attempts)
	if errors.Is(err, sql.ErrNoRows) {
		rlog.Warn(fmt.Sprintf("Job %s not found. Skipping.", event.JobID))
		return nil
	}
	if err != nil {
		return err
	}
	if job.Status == "COMPLETED" || job.Status == "SUPERSEDED" {
		return nil
	}

	var latestGeneration *int
	var lg int
	err = tx.QueryRow(ctx,
		`SELECT "latestGeneration" FROM "score_calc_state" WHERE "eventId" = $1`, event.EventID,
	).Scan(&lg)
	if err == nil {
		latestGeneration = &lg
	} else if !errors.Is(err, sql.ErrNoRows) {
		return err
	}

	if latestGeneration != nil && event.Generation < *latestGeneration {
		_, err = tx.Exec(ctx, `UPDATE "score_calc_task" SET status = 'SUPERSEDED' WHERE id = $1`, event.JobID)
		if err != nil {
			return err
		}
		return tx.Commit()
	}

	now := time.Now().UTC()
	if _, err := tx.Exec(ctx,
		`UPDATE "score_calc_task" SET status = 'FAILED', "lastError" = $1, "completedAt" = $2 WHERE id = $3`,
		event.ErrorMessage, now, event.JobID,
	); err != nil {
		return err
	}

	var userIDs []string
	if len(job.RequestedUserIDs) > 0 {
		_ = json.Unmarshal(job.RequestedUserIDs, &userIDs)
	}
	if err := tx.Commit(); err != nil {
		return err
	}

	if _, err := ScoreCalcStatusTopic.Publish(ctx, ScoreCalcStatusEvent{
		EventID: event.EventID, UserIDs: userIDs,
		Status: "failed", Timestamp: time.Now().UTC().Format(time.RFC3339Nano),
	}); err != nil {
		return err
	}
	return nil
}
